src/Symfony/Bundle/FrameworkBundle/Routing/Router.php

```diff
@@ -48,14 +48,12 @@ public function __construct(ContainerInterface $container, $resource, array $opt
     /**
      * {@inheritdoc}
      */
-    public function getRouteCollection()
+    protected function loadRouteCollection()
     {
-        if (null === $this->collection) {
-            $this->collection = $this->container->get('routing.loader')->load($this->resource, $this->options['resource_type']);
-            $this->resolveParameters($this->collection);
-        }
+        $collection = $this->container->get('routing.loader')->load($this->resource, $this->options['resource_type']);
+        $this->resolveParameters($collection);
```

Comment:
if any parameter got resolved, you need to add routing resources so that changes to the container invalidates the route collection. Otherwise, you will not refresh the RouteCollection properly, which would be a BC break compared to the non-cached behavior